### ISSUE TITLE
Cannot start nfs automount

### ISSUE THREAD

**Issue by EAlthuis** [vwal/nfs_automount#10]
On a Dell Vostro with newly installed Ubuntu 15.04, I followed the automount procedure precisely, when trying to start I got this message:

ea@vostro:~$ sudo start nfs-automount
[sudo] password for ea: 
start: Unable to connect to Upstart: Failed to connect to socket /com/ubuntu/upstart: Connection refused
ea@vostro:~$ 

I have no idea what to do next.
However I also understand that Ubuntu has switched from upstart to systemd, so I tried to "apt-get install upstart"  and got a message that upstart is installed already.

Any help to correct the problem would be greatly appreciated.

**Comment by zer0dayxploitz** [comment]
You need to convert the upstart service into a systemd service.  Create a file called nfs_automount.service in /lib/systemd/system/.  The contents of this file should look like this:

[Unit]
Description=Automatically mount, unmount, and remount NFS shares.
After=syslog.target
After=network.target

[Service]
Type=simple
Restart=on-failure
ExecStart=/usr/local/bin/nfs_automount

[Install]
WantedBy=multi-user.target

Then enable the new service with systemctl enable nfs_automount 
Next start the service with systemctl start nfs_automount

**Comment by EAlthuis** [comment]
Thank you for this information.

I had a number of other problems on the laptop that did not get 
resolved, so I removed all of my important data and started with Ubuntu 
12.04LTS and rebuilt from there, I am now running 14.04.2LTS and 
everything works as it should.

I still use upstart.  I will keep the info you provided, as I am sur I 
will need it in the future.

EA

**Comment by lmcogs** [comment]
Works lovely here

**Comment by EAlthuis** [comment]
This problem has been resolved, all computers now run 16.04LTS and use systemd as indicated above to start and maintain the NFS connection.

Thanks to all who helped mwe

**Comment by twtonicr** [comment]
I love this script, but it's been causing problems during machine shutdown with ubuntu hanging on a timeout.

I have nfs_automount working via systemd as advised above and it was originally working fine. 
I suspect this is what's happening during shutdown, instead of the service stopping, it restarts and times out endlessly waiting for the network.

Has something changed in the way systemd functions which now means something has to be added to nfs_automount.service in /lib/systemd/system/ ?

**Comment by EAlthuis** [comment]
I have found that the shutdown problem only occurs on the client systems.  I therefore shut down the clients before the server, and the problem is solved for me.  I am running Ubuntu 16.04LTS on all but one where I have Ubuntu-unity7 18.04LTS.
I have no knowledge of any changes to systemd.

**Comment by twtonicr** [comment]
I agree, a manual umount before shutdown is a solid workaround - Though I prefer the clients to handle this automatically as unskilled users want to dual-boot Ubuntu and Windows.

Possible solution:
I've added _netdev to the NFS options in /etc/nfs_automount.conf and in initial tests it seems to solve the problem.
We used it years ago but I though that _netdev was redundant now as systemd can automatically detect a network mount, but reading here it appears this is true only for NFS andd NFS4 mount types.
http://codingberg.com/linux/systemd_when_to_use_netdev_mount_option

We are using NFS3, and this might be the problem for our clients.

**Comment by EAlthuis** [comment]
As I am the owner and only user of NFS using 4 computers, the server and three others, all of which have unlimited access to the files on the server whenever it is on.
At this point I have no further problems, and insufficient knowledge of systemd to be able to either comment or provide suggestions.

**Comment by twtonicr** [comment]
Thanks EAlthuis, I'm grateful for your direct replies.
I'm updating here for others who might (like me) arrive at this thread when seeking answers to a problem.

Adding _netdev did not improve things.

I've done much research and this is a deeper problem with linux.  It's caused during shutdown by the fact that the network service gets killed before all the NFS files are unmounted, resulting in an endless timeout loop which prevents the system from reaching the shutdown target.

There are a number of workarounds, using a soft mount option apparently works but has a small but non-zero risk of silent data corruption.

The most promising results I've found are from delaying the stopping of the network service.
On my 16.04 clients, I've tried a solution from this bug report to modify the nfs-config.service:
https://bugs.launchpad.net/ubuntu/+source/nfs-utils/+bug/1577120 

This involves changing the line:
DEFAULT:
After=local-fs.target
NEW
After=local-fs.target remote-fs.target NetworkManager.service

In initial tests this is very promising. I haven't seen any shutdown hang and the umounts seem to be completing successfully.

As an aside, re zer0dayxploitz's great advice re systemd above, to follow convention the nfs_automount.service file should be created  in  /etc/systemd/system/
/usr/lib/systemd/system/: units provided by installed packages
/etc/systemd/system/: units installed by the system administrator

I hope this helps anyone having the same problem.
